server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessageListMethod.java

```diff
@@ -169,24 +169,35 @@ private long limit(Optional<Integer> limit) {
         return SortToComparatorConvertor.comparatorFor(messageListRequest.getSort());
     }
 
-    private ImmutableSet<MailboxPath> listRequestedMailboxes(GetMessageListRequest messageListRequest, List<MailboxPath> mailboxPaths, MailboxSession session) {
-        ImmutableSet<MailboxPath> mailboxPathSet = ImmutableSet.copyOf(mailboxPaths);
+    private List<MailboxPath> listRequestedMailboxes(GetMessageListRequest messageListRequest, List<MailboxPath> mailboxPaths, MailboxSession session) {
         return messageListRequest.getFilter()
-            .filter(FilterCondition.class::isInstance)
-            .map(FilterCondition.class::cast)
-            .map(FilterCondition::getInMailboxes)
-            .map(Throwing.function(mailboxIds -> mailboxIdsToMailboxPaths(mailboxIds, session)))
-            .map(requestedMailboxPaths -> Sets.intersection(requestedMailboxPaths, mailboxPathSet).immutableCopy())
-            .orElse(mailboxPathSet);
+                .filter(FilterCondition.class::isInstance)
+                .map(FilterCondition.class::cast)
+                .map(filterCondition -> filterMailboxPaths(mailboxPaths, session, filterCondition))
+                .orElse(mailboxPaths);
+    }
+
+    private List<MailboxPath> filterMailboxPaths(List<MailboxPath> mailboxPaths, MailboxSession session, FilterCondition filterCondition) {
+        Predicate<MailboxPath> inMailboxesPredicate = filterCondition.getInMailboxes()
+                .map(list -> mailboxIdsToMailboxPaths(list, session))
```

Comment:
flatmap mixing lists and optionals in not possible in Java (and I want to keep a list here)